### ISSUE TITLE
Define and manage sub-library integration more centrally, minimizing redundancies

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#6459]
So far there is high potential for inconsistencies w.r.t. sub-library/module integration, in particular when new ones are added. This is due to the rather intricate way sub-libraries are integrated in the crypto lib with the various meta files that need to be kept in sync. See for instance the section 'Adding new libraries' in `crypto/err/README`.
Integration is supported by various Perl scripts, yet they have their complexities in themselves and have non-trivial interactions also with the central `Makefile`.

**Comment by DDvO** [comment]
AFAICS, this is the list of files per LIB that need to be kept in sync:
```
Configure
util/libcrypto.num
crypto/err/openssl.txt
crypto/err/openssl.ec
crypto/err/err_all.c
crypto/err/err.c
include/openssl/LIB.h
include/openssl/LIBerr.h
include/openssl/err.h
crypto/LIB/build.info
crypto/LIB/LIB_err.c
```

**Comment by richsalz** [comment]
Some of those files are generated.  I am hoping we can have a single "list of dirs in crypto" and do all the right thing from that list.  We'll see.  That will then determine the requirements and ability to check for the various LIB files, for example.

**Comment by DDvO** [comment]
Also `crypto/build.info` needs to be extended, plus possibly `apps/build.info`.

In addition, `INSTALL `may need to be (manually) adapted, at least if the `no-LIB` option is offered.

**Comment by levitte** [comment]
- Configure

Hmm?  The only information that remains in there is about being able to disable chosen sub-modules, which is in no way mandatory...  or have I forgotten something?

- util/libcrypto.num
- crypto/err/openssl.txt
- include/openssl/LIBerr.h
- crypto/LIB/LIB_err.c

Automatically generated

- crypto/err/openssl.ec
- crypto/err/err_all.c
- crypto/err/err.c

True, and the whole error system needs an overhaul.  @richsalz, wasn't that something you were passionate about at some point?

Remains these:

- include/openssl/LIB.h
- include/openssl/err.h
- crypto/LIB/build.info

**Comment by DDvO** [comment]
Yes, w.r.t. `Configure` the situation has improved since last summer, but in case a module can be disabled this information, plus any related dependencies, must still be stated there.

I still had trouble today getting the auto-generated `include/openssl/LIBerr.h` and `crypto/LIB/LIB_err.c` bootstrapped at least using `make errors` and the like, or how best to do this?

Concerning the remaining ones you mentioned:

* `include/openssl/LIB.h` is of course no problem, since it is usually needed anyway. Same with any `crypto/LIB/LIB_int.h`, or should one better put internal headers in `crypto/include/internal/`?
* The `build.info` is needed not only in `crypto/LIB/`, but also `crypto/build.info` (and potentially `apps/build.info`) needs to be adapted. I think it would be best if these files were taken as master input for generating/updating the other meta files, in particular the error-related files.

**Comment by nhorman** [comment]
I'm not sure I see clearly the enumerated set of actions being requested.  Can someone state clearly for me:

1) When we state sub-library/module, what is being referred to?  Are we talking about external libraries (like zlib, etc), or providers (which seems like is not the case here, but we refer to providers as modules on occasion)
2) What the the centralization being asked for? Is the desire when new external libraries are added to the build, to have that integration take place in one single location (as opposed to a set of alternate locations throughout the tree)?  That seems like the case, but since generated files are being referred to above, Its not quite clear

**Comment by t8m** [comment]
IMO this is a "pink nirvana" kind of thing for future thinking.

**Comment by nhorman** [comment]
ok

Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
